Notes: do not prepend post titles to reply posts

jeherve opened the pull request:
## Proposed changes:

We currently force the addition of the post title to notes, so that title is included before the post content if it exists.

This does not make sense when the note includes a Federated Reply block. In those situations, you would expect the beginning of the note to be a @mention of the person you are replying to, not the post title.

> [!NOTE]
> This PR is also an opportunity to add an extra parameter to the existing `activitypub_object_content_template`, so it's easier for folks to customize their content template based off the type of the post.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Write and publish 2 posts:
    * One that includes a Federated reply block in reply to a random post on the Fediverse.
    * One that does not.
* Both posts should have a post title, but should have very content in the post so they're automatically classified as "Notes".
* View the posts on the frontend, and add `/activitypub` to the end of the post URL
    * The post title should only be prepended to the post that does not include the reply block.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

<!-- Add a changelog message here -->

</details>

jeherve committed:
Post content template: add object type to existing filter

It will allow third-parties to more easily customize the post content template based on the object type.

jeherve committed:
Do not prepend title to note content when post includes a reply

It does not make sense for a post to include a title when the beginning of the message should be a mention of the person you are replying to.

jeherve committed:
Add changelog

pfefferle commented on includes/transformer/class-post.php:
```diff
@@ -894,11 +894,15 @@ protected function get_post_content_template() {
-			if ( 'Note' === $this->get_type() ) {
+			// If the post is a note and not a reply, force the inclusion of the post title.
+			$is_reply = null !== $this->get_in_reply_to();
+
```
@jeherve Should we also check for `$this->get_mentions()` (@-replies) or do you think replies are fine for now?

jeherve replied:
I worry this would catch longer posts, that would only happen to mention Fediverse users like we mention someone by their name and link to their blog or Twitter account today. 

Here is an example:
https://herve.bzh/where-did-all-the-wordpress-editors-go/

jeherve replied:
After our discussion, I added this in since it's unlikely note posts would be longer posts.

d39a915facdcbd85b12068436149c137e3eb7aaf

pfefferle committed:
Merge branch 'trunk' into fix/note-title-detection

obenland commented on includes/transformer/class-post.php:
```diff
@@ -894,11 +894,15 @@ protected function get_post_content_template() {
+			// If the post is a note and not a reply, force the inclusion of the post title.
+			$is_reply = null !== $this->get_in_reply_to();
+
+			if ( 'Note' === $type && ! $is_reply ) {
```
```suggestion
			if ( 'Note' === $type && empty( $this->get_in_reply_to() ) ) {
```
Would this work?

It would be great to have a unit test for this

jeherve replied:
Works for me! c289589abbde6696e917b4ec8d92256c94a18e80

jeherve committed:
Add comment to show that the last parameter was added later

obenland committed:
simplify check

See https://github.com/Automattic/wordpress-activitypub/pull/2419#discussion_r2495288741

jeherve committed:
Add unit test to cover new scenario

Also refactor existing tests so they're all in one test with multiple scenarios.

jeherve committed:
Do not add title when the post includes mentions

See https://github.com/Automattic/wordpress-activitypub/pull/2419#discussion_r2495069961

jeherve committed:
Add test for new case (mentions)

jeherve committed:
Merge branch 'fix/note-title-detection' of github.com:Automattic/wordpress-activitypub into fix/note-title-detection

jeherve commented on includes/transformer/class-post.php:
```diff
@@ -915,8 +919,9 @@ protected function get_post_content_template() {
 		 *
 		 * @param string   $template  The template string containing shortcodes.
 		 * @param \WP_Post $item The WordPress post object being transformed.
+		 * @param string   $type ActivityStreams 2.0 Object-Type for the post.
```
Done in cac66546b3668f96465512bda51048883a2f65b4

jeherve commented:
> It would be great to have a unit test for this

Sure thing. I worked on that in bfdfbc8d302cefb6d513cb2017ce331c2daa6862

jeherve committed:
Fix spacing

pfefferle commented on includes/transformer/class-post.php:
```diff
@@ -894,11 +894,20 @@ protected function get_post_content_template() {
+			 * force the inclusion of the post title.
+			 */
+			if (
+				'Note' === $type
+				&& empty( $this->get_in_reply_to() )
+				&& empty( $this->get_mentions() )
+			) {
```
Shouldn't it be either or?

```suggestion
				'Note' === $type &&
				( empty( $this->get_in_reply_to() ) || 
				empty( $this->get_mentions() ) )
			) {
```

pfefferle replied:
Oh no, it checks for empty! You are right!!! Forget my comment 🫣

pfefferle reviewed (approved):
Thanks @jeherve ☺️
